### PR TITLE
Replace hashlib.ripemd160 with pythonic implementation

### LINKED ISSUE #118: Unsupported hash type ripemd160
Tested on OpenSSL 3.0.2, Python 3.10, Linux Ubuntu 22.04 LTS.

Hashlib raises ValueError: unsupported hash type ripemd160, in ripemd160_sha256 function in crypto.py. Found the reason for the issue [here](https://stackoverflow.com/questions/72409563/unsupported-hash-type-ripemd160-with-hashlib-in-python).

>Hashlib uses OpenSSL for ripemd160 and apparently OpenSSL disabled some older crypto algos around version 3.0 in November 2021. All the functions are still there but require manual enabling. See [issue 16994 of OpenSSL github project](https://github.com/openssl/openssl/issues/16994) for details.

The stackoverflow answer details how to enable legacy crypto algorithms, but I think a default solution is needed to use ripemd160.

### PR DESCRIPTION
**Pull request by yashasvi-ranawat**
Closes #118

### CODE REVIEW

**Comment by yashasvi-ranawat** [comment]
the pythonic implementation is 2 orders magnitude slower. Will only use it as fallback, in the next commit

**Comment by yashasvi-ranawat** [comment]
In the present application of the function, where it is only used to generate P2PKH locking script (and cashaddress), I think this is okay.

However, if in the future BitCash implements P2SH wallets, then the wallet can leak what underlying implementation it uses when asked to verify a script. Here, of course, switching to P2SH32 would be a solution, which also follows the general recommendation against using ripemd160.

File: bitcash/crypto.py [hunk 1]
```diff
@@ -16,7 +17,10 @@ def double_sha256_checksum(bytestr):
 
 
 def ripemd160_sha256(bytestr):
-    return new("ripemd160", sha256(bytestr)).digest()
+    try:
+        return new("ripemd160", sha256(bytestr)).digest()
+    except Exception:
```
Comment (merc1er) [inline bitcash/crypto.py:22]: Could this be narrowed down to the very exception(s) it throws in case OpenSSL fails to produce a `ripemd160`?

**Review by merc1er — CHANGES_REQUESTED** [review]
Thanks for this! I will set up different environments in order to test that.

Could you please add unit tests to all individual functions as well as for `ripemd160_sha256`?

**Comment by merc1er** [comment]
> Here, of course, switching to P2SH32 would be a solution, which also follows the general recommendation against using ripemd160.

Sorry, I am not familiar with `P2SH32` and the upcoming network upgrade. Is `P2SH32` meant to replace `P2SH` or work along it? And does that meant ripemd160 will no longer be necessary?

**Comment by yashasvi-ranawat** [comment]
> > Here, of course, switching to P2SH32 would be a solution, which also follows the general recommendation against using ripemd160.
> 
> Sorry, I am not familiar with `P2SH32` and the upcoming network upgrade. Is `P2SH32` meant to replace `P2SH` or work along it? And does that meant ripemd160 will no longer be necessary?

It is meant to be go along with P2SH (since a lot of contracts have used it, and a lot of wallets have it implemented). The primary idea is to limit hash collision of complex scripts, which is a possibility with P2SH.

Although, ripemd160 will still be relevant with P2PKH and P2SH. I haven't completely followed the discussion either, but [it is heavily recommended for the P2SH scripts to switch to P2SH32](https://bitcoincashresearch.org/t/p2sh32-a-long-term-solution-for-80-bit-p2sh-collision-attacks/750).

File: bitcash/crypto.py (hunk 1 above)
Comment (yashasvi-ranawat) [reply bitcash/crypto.py]: done

**Comment by yashasvi-ranawat** [comment]
> Could you please add unit tests to all individual functions as well as for `ripemd160_sha256`?

I monkeypatched tests for pythonic ripemd160_sha256. It should test it on any environment.

**Comment by yashasvi-ranawat** [comment]
>Could you please add unit tests to all individual functions as well as for ripemd160_sha256?

Correct monkeypatching implemented to test pythonic ripemd160. should test it regardless of environment.

**Comment by merc1er** [comment]
Is this ready for final review @yashasvi-ranawat?

**Review by merc1er — APPROVED** [review]
Thank you for this!

**Pull request merged**